File: server/src/utils/env.ts
import fs from "fs";

/**
 * Reads environment variables, prioritizing `_FILE` variants.
 * If the `_FILE` variant exists, it reads the value from the file.
 * Otherwise, it falls back to the standard environment variable.
 * @example
 * ```ts
 * // if the environment variable MY_VAR_FILE exists, it will read from that file
 * // otherwise, it will return the value of MY_VAR
 * const myVar = getEnv("MY_VAR");
 * ```
 *
 * @param key - The environment variable key.
 * @returns The value of the environment variable or file content.
 */
export const getEnv = (key: string): string | undefined => {
  const fileKey = `${key}_FILE`;
  if (process.env[fileKey]) {
    try {
      return fs.readFileSync(process.env[fileKey], "utf8").trim();
    } catch (error) {
      console.error(`Error reading file for ${fileKey}:`, error);
    }
  }
  return process.env[key];
};

File: email-service/utils/env.go
package utils

import (
	"log"
	"os"
)

// @INFO GetEnv reads environment variables, prioritizing `_FILE` variants.
// If the `_FILE` variant exists, it reads the value from the file.
// Otherwise, it falls back to the standard environment variable.
func GetEnv(key string) string {
	fileKey := key + "_FILE"
	if filePath, exists := os.LookupEnv(fileKey); exists {
		content, err := os.ReadFile(filePath)
		if err != nil {
			log.Printf("Error reading file for %s: %v", fileKey, err)
			return ""
		}
		return string(content)
	}
	return os.Getenv(key)
}
